plugins/fabric8-end2end-flow/src/main/java/com/redhat/che/end2end/End2EndFlowService.java

```diff
@@ -0,0 +1,217 @@
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.Status;
+import javax.ws.rs.core.StreamingOutput;
+import org.eclipse.che.api.core.rest.Service;
+import org.eclipse.che.commons.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Singleton
+@Path("fabric8-end2end")
+public class End2EndFlowService extends Service {
+  private static final Logger LOG = LoggerFactory.getLogger(End2EndFlowService.class);
+
+  private final String reCaptchaSiteKey;
+  private final String reCaptchaSecretKey;
+
+  private final Map<String, Function<String, String>> staticFilesFilters;
+
+  @Inject
+  public End2EndFlowService(
+      @Nullable @Named("che.fabric8.end2end.protect.site_key") String siteKey,
+      @Nullable @Named("che.fabric8.end2end.protect.secret_key") String secretKey) {
+    this.reCaptchaSiteKey = siteKey;
+    this.reCaptchaSecretKey = secretKey;
+    this.staticFilesFilters = new HashMap<>();
+    String siteKeyDecl;
+    if (siteKey != null && !siteKey.isEmpty()) {
+      siteKeyDecl = "var siteKey='" + siteKey + "';";
+    } else {
+      siteKeyDecl = "var siteKey='';";
+      LOG.warn("No ReCaptcha site key was provided. ReCaptcha user verification is disabled !");
+    }
+    staticFilesFilters.put(
+        "files/provision.html", line -> line.replace("const siteKey;", siteKeyDecl));
+  }
+
+  @GET
+  @Path("site-key")
+  public String siteKey() {
+    return reCaptchaSiteKey == null ? "" : reCaptchaSiteKey;
+  }
+
+  @POST
+  @Path("verify")
+  @Consumes("text/plain")
+  @Produces(APPLICATION_JSON)
+  public Response verifyReCaptchaToken(String token) {
+    int responseCode = 500;
+    try {
+      URL obj = new URL("https://www.google.com/recaptcha/api/siteverify");
+      HttpsURLConnection con = (HttpsURLConnection) obj.openConnection();
+
+      con.setRequestMethod("POST");
+
+      String urlParameters = "secret=" + reCaptchaSecretKey + "&response=" + token;
+
+      con.setDoOutput(true);
+      try (DataOutputStream wr = new DataOutputStream(con.getOutputStream())) {
+        wr.writeBytes(urlParameters);
+        wr.flush();
+      }
+
+      responseCode = con.getResponseCode();
+
+      if (responseCode == 200) {
+        StringBuffer response = new StringBuffer();
+        try (BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()))) {
+          String inputLine;
+          while ((inputLine = in.readLine()) != null) {
+            response.append(inputLine);
+          }
+        }
+        return Response.ok(response.toString()).build();
+      } else {
+        LOG.error(
+            "reCaptcha verification failed with the fllowing response code: {}", responseCode);
```

Comment:
Might be helpful to add response body to the log in a case some unexpected error is received